lib/IPCPipeSend.cpp

```diff
@@ -0,0 +1,91 @@
+/** @file IPCPipeSend.cpp
+ *  @author Bao Vu (bao.vu@unikie.com)
+ *  @brief
+ */
+
+#include "IPCPipeSend.h"
+
+IPCPipeSend::~IPCPipeSend()
+{
+	if (this->pd > 0)
+		close(this->pd);
+	unlink(this->opts.server_name.c_str());
+}
+
+void IPCPipeSend::init()
+{
+	// remove old pipe (if any)
+	remove(this->opts.server_name.c_str());
+	// create a fifo
+	if (mkfifo(this->opts.server_name.c_str(), 0660) != 0)
+		throw std::runtime_error("ERROR: " + this->opts.server_name + ": " + strerror(errno));
+	// open the pipe with O_NONBLOCK to get EAGAIN
+	this->pd = open(this->opts.server_name.c_str(), O_RDWR | O_NONBLOCK);
+	if (this->pd == -1)
+		throw std::runtime_error("ERROR: " + this->opts.server_name + ": " + strerror(errno));
+	else
+		std::cout << "[" << this->opts.server_name << "] is opened." << std::endl;
+
+
+	this->open_file();
+	this->info.file_size = this->get_file_size();
+	if (this->info.file_size == 0)
+		throw std::runtime_error("ERROR: File size = 0.");
+	this->timer.update_all();
+
+}
+
+void IPCPipeSend::transfer()
+{
+	std::vector<char> buffer(this->p_msgsize);
+
+	std::cout << "Sending..." << std::endl;
+	while (!(this->fs.eof()))
+	{
+		this->fs.read(buffer.data(), this->p_msgsize);
+		if (this->fs.bad()) // check read/writing error on i/o operation
+			throw std::runtime_error("ERROR: istream::read().");
+
+		this->info.read_bytes = this->fs.gcount();
```

Comment:
read_bytes, send_bytes  is used only in 'transfer', so in my opinion, code will be more clear and safer if it will be local variable, not class attribute. The same for PipeReceive